.github/workflows/reusable-docs.yml

```diff
@@ -18,6 +18,8 @@ jobs:
     timeout-minutes: 60
     steps:
     - uses: actions/checkout@v3
+      with:
+        fetch-depth: 0
```

Comment:
We _should_ only need the history of both branches up to and including the common merge-base. However, unfortunately that's non-trivial to retrieve without hacky, non-trivial and potentially unreliable workarounds. See actions/checkout#552 for more discussion of this. 

While ideally I'd like to avoid this, in the end its an extra 1 minute (30%) to the build time of one of our docs jobs, which brings it to being about as long as the RTD preview build (≈4 minutes each), which is what most people probably care about anyway, so in practice the impact is limited. We could potentially explore workarounds here, but that's probably best left for a followup PR.

Reply:
And I recall someone who was unhappy about the length of the RTD build ;)

The CPython history is huge: 33 years and 120k commits. Could we limit depth to 100 commits? 500? 1,000?

Are any of the options from https://github.blog/2020-12-21-get-up-to-speed-with-partial-clone-and-shallow-clone/ possible, such as blobless clone?

Reply:
> And I recall someone who was unhappy about the length of the RTD build ;)

Hehe, yeah, though that was because it was a long pole on PR checks and the main thing I wanted to look at, so making the regular build run at the same speed is one "solution" to that :joy: 

Mostly, I was just too tired to think more about this after putting way more time and energy into this than I expected yesterday, including debugging issues related to not fetching the the refs we needed here, because it requires such a long iteration cycle (>5 minutes per run) with minimal feedback and no interactivity.

Unfortunately, given:

* There's around 500 commits per month just on `main`
* People often work on branches that are many months or even years old
* If the fork point was earlier than the shallow depth, the CI job will cryptically fail
* AFAIK, because it doesn't change the original fork point, merging with `main` (AFAIK) won't fix this, only rebasing the branch and rewriting history

We'd have to set a pretty high shallow clone depth, of at least >10 000+ to be reasonably unlikely to run into this, and we might still hit it on a long tail of PRs, plus we might still need other workarounds in the code anyway. So it seems we're probably better off either leaving it, or trying to implement the workarounds at the issue linked I cited above. I can take a crack at the latter, especially if I can repro the shallow clone behavior locally, which would make testing, debugging and iterating _way_ easier.

Yeah, a blobless or treeless clone would be a great help here and a big improvement over a "dumb" shallow clone, but unfortunately despite GitHub itself writing that blog post several years ago, there's inexplicably still no support for anything but a bare-bones shallow clone in the official `checkout` action (which, apparently, is required for reusable workflows, which this in fact is); see, e.g. actions/checkout#1152 and many other issues there, so we'll have to hack something together on our own.

Reply:
To make a very long and painful story short, after much testing and experimentation, I managed to adapt a solution from https://github.com/actions/checkout/issues/520#issuecomment-1167205721 to work here for fetching just enough history to compare the two branches, which adds less than a second to the build process, instead of over a minute. In total, at a rough average of ≈50-ish runs per day of this job (out of 125 workflow runs total), it'll pay back the hours I spent on it, at least in CI time, in under a week.

Also, in testing this, I noticed another pre-existing issue present with the previous version: because the workflow runs on the PR ref _merged_ with the base ref, the lines being reported would be out of sync whenever there were upstream changes, since the GitHub annotations and the Files view are based on the line numbers from the PR head ref, not the merge ref. Naturally, I fixed that as well by making the main docs job run on the PR head ref (which also made the shallow checkout more reliable).